src/Microsoft.Azure.WebJobs.Host/Bindings/QueueBindingRules.cs

```diff
@@ -0,0 +1,217 @@
+    // This is fundemantentally an IAsyncCollector<IStorageQueueMessage>
+    internal class QueueBindingRules
+    {
+        private readonly IStorageAccountProvider _accountProvider;
+        private readonly IContextGetter<IMessageEnqueuedWatcher> _messageEnqueuedWatcherGetter;
+
+        public QueueBindingRules(IStorageAccountProvider accountProvider, IContextGetter<IMessageEnqueuedWatcher> messageEnqueuedWatcherGetter)
+        {
+            _accountProvider = accountProvider;
+            _messageEnqueuedWatcherGetter = messageEnqueuedWatcherGetter;
+        }
+
+        public IBindingProvider BuildQueueRules(
+            INameResolver nameResolver,
+            IConverterManager converterManager)
+        {
+            // IStorageQueueMessage is the core testing interface 
+            converterManager.AddConverter<byte[], IStorageQueueMessage, QueueAttribute>(ConvertByteArrayCloudQueueMessage);
+            converterManager.AddConverter<string, IStorageQueueMessage, QueueAttribute>(ConvertString2CloudQueueMessage);
+            converterManager.AddConverter<IStorageQueueMessage, string>(ConvertCloudQueueMessage2String);
+            converterManager.AddConverter<IStorageQueueMessage, byte[]>(ConvertCloudQueueMessage2ByteArray);
+
+            converterManager.AddConverter<CloudQueueMessage, IStorageQueueMessage>(ConvertReal2CloudQueueMessage);
+
+            var bf = new BindingFactory(nameResolver, converterManager);
+
+            var ruleQueueOutput = bf.BindToAsyncCollector<QueueAttribute, IStorageQueueMessage>(BuildFromQueueAttribute, ToParameterDescriptorForCollector, FixerUpper);
+            var ruleQueueClient = bf.BindToExactAsyncType<QueueAttribute, IStorageQueue>(BuildClientFromQueueAttributeAsync, ToParameterDescriptorForCollector, FixerUpper);
+            var ruleQueueClient2 = bf.BindToExactAsyncType<QueueAttribute, CloudQueue>(BuildRealClientFromQueueAttributeAsync, ToParameterDescriptorForCollector, FixerUpper);
+            var queueRules = new GenericCompositeBindingProvider<QueueAttribute>(
+                (attr) => ValidateQueueAttribute(attr, nameResolver),
+                ruleQueueClient, ruleQueueClient2, ruleQueueOutput);
+
+            return queueRules;
+        }
+
+        // Hook to apply hacky rules all at once. 
+        private async Task<QueueAttribute> FixerUpper(QueueAttribute attrResolved, ParameterInfo parameter, INameResolver nameResolver)
+        {
+            // Look for [Storage] attribute and squerrel aover 
+            IStorageAccount account = await _accountProvider.GetStorageAccountAsync(parameter, CancellationToken.None, nameResolver);
+            StorageClientFactoryContext clientFactoryContext = new StorageClientFactoryContext
+            {
+                Parameter = parameter
+            };
+            IStorageQueueClient client = account.CreateQueueClient(clientFactoryContext);
+
+            // Create the queue if needed?
```

Comment:
Previously the queue was created on demand the first time a message was added. It looks like you're doing this eagerly now at bind/startup time?

I think that's the more desirable behavior - i.e. if I have 20 queue functions pointing to different queues but I only ever use 1, I don't want all 20 queues created. I think that will be true of a lot of other bindings - e.g. for expensive resources like DocumentDB, I don't want to create tables until they're actually used.